[TECH] Remplacer l'utilisation de la propriété de configuration endSessionUrl (PIX-11821)

lego-technix opened the pull request:
## :unicorn: Problème

Avec l'ajout de la librairie _openid-client_ avec #8028, nous n'avons plus besoin de la variable d'environnement `${SSO}_END_SESSION_URL`.

Un nettoyage avait déjà été fait précédemment avec #8136 (et pour cause d'une régression la restauration de certaines propriétés avait été faite avec #8214) et #8458.

## :robot: Proposition

* Remplacer les dernières utilisations de `OidcAuthenticationService#endSessionUrl` (et de `hasLogoutUrl`) par l'utilisation d'une nouvelle propriété générique : `shouldCloseSession`

* Dans les usecases,  appeler `oidcAuthenticationService#saveIdToken` uniquement si `oidcAuthenticationService#shouldCloseSession`

* Et aussi dans la même PR ajouter la colonne `shouldCloseSession` correspondante dans la table `oidc-providers`.

## :rainbow: Remarques

Il reste des variables `endSessionUrl` dans le code, mais elles sont récupérées à partir de l'OpenID Configuration endpoint et non pas à partir de la configuration.

## :100: Pour tester

### En RA

Pour l'instant on ne peut pas tester en RA.

### En local

1. Réaliser des tests de non-régression avec tous les SSO, en vérifiant particulièrement que le logout spécifique de chacun des SSO a bien le même comportement qu'avant cette PR.
2. Réaliser la validation de la migration additionnelle de la DB : 
   1. Exécuter la commande `npm run db:migrate` en local
   2. Vérifier que la table `oidc-providers` a bien été modifiée avec l'ajout d'une colonne `shouldCloseSession`
   3. Exécuter la commande `npm run db:rollback:latest`et vérifier que la table `oidc-providers` ne contient plus la nouvelle colonne `shouldCloseSession`

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr8494.review.pix.fr)
  * [App (.org)](https://app-pr8494.review.pix.org)
  * [Orga (.fr)](https://orga-pr8494.review.pix.fr)
  * [Orga (.org)](https://orga-pr8494.review.pix.org)
  * [Certif (.fr)](https://certif-pr8494.review.pix.fr)
  * [Certif (.org)](https://certif-pr8494.review.pix.org)
  * [1D](https://1d-pr8494.review.pix.fr)
  * [Admin](https://admin-pr8494.review.pix.fr)
  * [API](https://api-pr8494.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr8494.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr8494/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr8494/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr8494/environment)

igorissen commented on api/db/migrations/20240325104111_add-doesLogout-and-doesCustomLogout-to-table-oidc-providers.js:
```diff
@@ -0,0 +1,19 @@
+const TABLE_NAME = 'oidc-providers';
+const COLUMN1_NAME = 'doesLogout';
+const COLUMN2_NAME = 'doesCustomLogout';
```
```suggestion
const COLUMN_NAME = 'shouldEndSession';
```

Je pense qu'une seule variable est nécessaire et peut être trouver un autre nom comme par exemple `shouldEndSession` qui porte sur le faite que l'on doit ou non clôturer la session du partenaire.

On peut faire une mob review avec @1024pix/team-acces pour en discuter.

lego-technix replied:
Il a été décidé en équipe de choisir le nom `shouldCloseSession` qui permet d'indiquer qu'il s'agit d'une propriété générique non-proche du mot `endSession` utilisé pour la fonctionnalité de logout standard d'OIDC.

lego-technix committed:
feat(api): add migration to oidc-providers to add

shouldCloseSession column to oidc-providers table, to store 1 more needed
property.

lego-technix committed:
refactor(api): replace hasLogoutUrl by a generic

shouldCloseSession

lego-technix committed:
refactor(api): replace endSessionUrl by generic

shouldCloseSession to remove the dependency on endSessionUrl which is obtained
through the OpenID Configuration endpoint of the OIDC Provider.

lego-technix committed:
refactor(mon-pix): replace hasLogoutUrl by a generic

shouldCloseSession

lego-technix committed:
refactor(admin): replace hasLogoutUrl by a generic

shouldCloseSession

lego-technix committed:
refactor(mon-pix): replace useEndSession by generic

shouldCloseSession to remove the dependency on endSessionUrl which is obtained
through the OpenID Configuration endpoint of the OIDC Provider.

lego-technix committed:
refactor(api): more logical only call saveIdToken

when shouldCloseSession, which does not make it mandatory to implement
saveIdToken for OIDC Providers having no support for logout.

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

igorissen commented on api/src/authentication/domain/services/oidc-authentication-service.js:
```diff
@@ -141,7 +139,7 @@ class OidcAuthenticationService {
 
   async saveIdToken({ idToken, userId } = {}) {
-    if (!(this.endSessionUrl || this.hasLogoutUrl)) {
+    if (!this.shouldCloseSession) {
```
Si on garde ce code dans la méthode `saveIdToken`, alors je pense qu'il faudrait retirer toutes les modifications dans les usecases.

```javascript
let logoutUrlUUID;
  if (oidcAuthenticationService.shouldCloseSession) {
    logoutUrlUUID = await oidcAuthenticationService.saveIdToken({
      idToken: sessionContent.idToken,
      userId,
    });
  }
```

```javascript
    const logoutUrlUUID = await oidcAuthenticationService.saveIdToken({
      idToken: sessionContent.idToken,
      userId,
    });
```

lego-technix replied:
Voilà, la condition ` if (!this.shouldCloseSession)` dans `oidcAuthenticationService#saveIdToken` a été supprimée puisque cette condition est déjà présente dans les usecases.

igorissen reviewed (approved):
✅ sur Safari _Version 17.3.1 (19617.2.4.11.12)_ avec tous les partenaires

mariannebost commented:
Test de la migration OK

EmmanuelleBonnemay commented on api/tests/authentication/unit/domain/services/oidc-authentication-service_test.js:
```diff
@@ -241,13 +243,22 @@ describe('Unit | Domain | Services | oidc-authentication-service', function () {
   describe('#saveIdToken', function () {
     it('should return null', async function () {
```
```suggestion
it('should return an idToken matching uuidV4Regex', async function () {
```

EmmanuelleBonnemay reviewed (approved):
Lu et testé en local avec succès avec Firefox
